### PR TITLE
Feature 202 editweekplan

### LINKED ISSUE #202: As a guardian I would like to be able to edit weekplan details (name, picture) so that I can fix mistakes I made
**Is your feature request related to a problem? Please describe.**
Right now it is not possible to change the name or image of a weekplan

**Describe the solution you'd like**
I would like to somehow get in to an edit menu of sorts where I can change the image and name of the weekplan

**Describe alternatives you've considered**
no

**Additional context**
null

### PR DESCRIPTION
**Pull request by EsbenNedergaard**
This pull request is bound to issue #202, about being able to edit week plans.

### CODE REVIEW

File: android/app/src/main/java/io/flutter/plugins/GeneratedPluginRegistrant.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ */
+@Keep
+public final class GeneratedPluginRegistrant {
+  public static void registerWith(@NonNull FlutterEngine flutterEngine) {
+    ShimPluginRegistry shimPluginRegistry = new ShimPluginRegistry(flutterEngine);
+      io.flutter.plugins.flutter_plugin_android_lifecycle.FlutterAndroidLifecyclePlugin.registerWith(shimPluginRegistry.registrarFor("io.flutter.plugins.flutter_plugin_android_lifecycle.FlutterAndroidLifecyclePlugin"));
+    flutterEngine.getPlugins().add(new io.flutter.plugins.imagepicker.ImagePickerPlugin());
+    flutterEngine.getPlugins().add(new io.flutter.plugins.sharedpreferences.SharedPreferencesPlugin());
+  }
+}
```
Comment (joandrsn) [inline android/app/src/main/java/io/flutter/plugins/GeneratedPluginRegistrant.java]: Why does this file exist?
It was removed in commit b3db5711bc4347341407a3be294a1bed1968e2f7 (which has been merged into the `develop` branch).
Is it maybe a wrong merge on this branch?

File: ios/Runner/GeneratedPluginRegistrant.m [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+#endif
+
+@implementation GeneratedPluginRegistrant
+
++ (void)registerWithRegistry:(NSObject<FlutterPluginRegistry>*)registry {
+  [FLTImagePickerPlugin registerWithRegistrar:[registry registrarForPlugin:@"FLTImagePickerPlugin"]];
+  [FLTSharedPreferencesPlugin registerWithRegistrar:[registry registrarForPlugin:@"FLTSharedPreferencesPlugin"]];
+}
+
+@end
```
Comment (joandrsn) [inline ios/Runner/GeneratedPluginRegistrant.m]: Why does this file exist?
It was removed in commit b3db5711bc4347341407a3be294a1bed1968e2f7 (which has been merged into the `develop` branch).
Is it maybe a wrong merge on this branch?

File: lib/blocs/activity_bloc.dart [hunk 3]
```diff
@@ -1,9 +1,9 @@
-import 'package:rxdart/rxdart.dart';
-import 'package:weekplanner/blocs/bloc_base.dart';
-import 'package:api_client/models/activity_model.dart';
-import 'package:api_client/models/username_model.dart';
 import 'package:api_client/api/api.dart';
+import 'package:api_client/models/activity_model.dart';
 import 'package:api_client/models/enums/activity_state_enum.dart';
+import 'package:api_client/models/username_model.dart';
+import 'package:rxdart/rxdart.dart';
```
Comment (joandrsn) [inline lib/blocs/activity_bloc.dart]: Why do you reorder the imports?

File: lib/blocs/edit_weekplan_bloc.dart [hunk 4]
```diff
@@ -0,0 +1,47 @@
+import 'package:api_client/api/api.dart';
+import 'package:api_client/models/username_model.dart';
+import 'package:api_client/models/week_model.dart';
+import 'package:rxdart/rxdart.dart';
+import 'package:weekplanner/blocs/new_weekplan_bloc.dart';
+import 'package:weekplanner/blocs/weekplan_selector_bloc.dart';
+
+class EditWeekplanBloc extends NewWeekplanBloc {
+  /// This class is an extension of NewWeekplan bloc
```
Comment (Jakobzs) [inline lib/blocs/edit_weekplan_bloc.dart]: It is already known that the class extends NewWeekplanBloc. A comment is unnecessary.

File: lib/blocs/add_activity_bloc.dart [hunk 5]
```diff
@@ -1,10 +1,9 @@
-import 'package:weekplanner/blocs/bloc_base.dart';
-import 'package:rxdart/rxdart.dart';
 import 'package:flutter/material.dart';
+import 'package:rxdart/rxdart.dart';
+import 'package:weekplanner/blocs/bloc_base.dart';
 
 /// Bloc used for adding activities to a citizen's day
-class AddActivityBloc extends BlocBase{
-
+class AddActivityBloc extends BlocBase {
```
Comment (joandrsn) [inline lib/blocs/add_activity_bloc.dart]: Please exclude formatting changes to files which have no impact on the PR.

File: lib/blocs/edit_weekplan_bloc.dart [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+  EditWeekplanBloc(Api api) : super(api);
+
+  /// This method should always be called before using the bloc
```
Comment (Jakobzs) [inline lib/blocs/edit_weekplan_bloc.dart]: Besides detailing that it has to be called first, it'd be great to have some information on what the method does.

File: lib/blocs/auth_bloc.dart [hunk 7]
```diff
@@ -53,9 +53,9 @@ class AuthBloc extends BlocBase {
   void authenticateFromPopUp(String username, String password) {
     _api.account.login(username, password).listen((bool status) {
       if (status) {
-          _loginAttempt.add(status);
-          setMode(WeekplanMode.guardian);
-        }
+        _loginAttempt.add(status);
+        setMode(WeekplanMode.guardian);
+      }
```
Comment (joandrsn) [inline lib/blocs/auth_bloc.dart]: Formatting.

File: lib/blocs/edit_weekplan_bloc.dart [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+  void initializeEditBloc(UsernameModel user, WeekModel weekModel) {
+    super.initialize(user);
+    // We just take the values out of the week model and put into our sink
+    super.onTitleChanged.add(weekModel.name);
+    super.onYearChanged.add(weekModel.weekYear.toString());
+    super.onWeekNumberChanged.add(weekModel.weekNumber.toString());
+    super.onThumbnailChanged.add(weekModel.thumbnail);
+  }
+
+  /// This method allows one to edit
```
Comment (Jakobzs) [inline lib/blocs/edit_weekplan_bloc.dart]: One to edit what?

File: lib/blocs/pictogram_image_bloc.dart [hunk 9]
```diff
@@ -1,11 +1,12 @@
 import 'dart:async';
 import 'dart:collection';
+
+import 'package:api_client/api/api.dart';
+import 'package:api_client/models/pictogram_model.dart';
 import 'package:flutter/material.dart';
+import 'package:mutex/mutex.dart';
```
Comment (joandrsn) [inline lib/blocs/pictogram_image_bloc.dart]: Please undo reorder of imports.

File: lib/blocs/timer_bloc.dart [hunk 10]
```diff
@@ -101,7 +102,6 @@ class TimerBloc extends BlocBase {
                     c.remaining.inMilliseconds));
           });
         } else if (_activityModel.timer.paused) {
-
           _timerRunningStream.add(false);
```
Comment (joandrsn) [inline lib/blocs/timer_bloc.dart]: Formatting please.

File: lib/blocs/toolbar_bloc.dart [hunk 11]
```diff
@@ -260,16 +256,16 @@ class ToolbarBloc extends BlocBase {
...
-                // Timer makes it clicable again after 2 seconds.
-                Timer(const Duration(milliseconds: 2000), () {
-                  _clickable = true;
-                });
-              }
-            }
+                    if (_clickable) {
+                      _clickable = false;
+                      loginFromPopUp(context, _authBloc.loggedInUsername,
+                          passwordCtrl.value.text);
```
Comment (joandrsn) [inline lib/blocs/toolbar_bloc.dart]: Formatting.

File: lib/bootstrap.dart [hunk 12]
```diff
@@ -1,21 +1,22 @@
...
 import 'package:weekplanner/blocs/new_weekplan_bloc.dart';
 import 'package:weekplanner/blocs/pictogram_bloc.dart';
 import 'package:weekplanner/blocs/pictogram_image_bloc.dart';
 import 'package:weekplanner/blocs/settings_bloc.dart';
-import 'package:weekplanner/blocs/upload_from_gallery_bloc.dart';
 import 'package:weekplanner/blocs/timer_bloc.dart';
-import 'package:weekplanner/blocs/weekplan_selector_bloc.dart';
 import 'package:weekplanner/blocs/toolbar_bloc.dart';
-import 'package:weekplanner/blocs/add_activity_bloc.dart';
+import 'package:weekplanner/blocs/upload_from_gallery_bloc.dart';
```
Comment (joandrsn) [inline lib/bootstrap.dart]: Dont reorder imports. 
Git sees this as a linechange, when it is indeed a new line.

File: lib/screens/choose_citizen_screen.dart [hunk 13]
```diff
@@ -56,10 +55,10 @@ class ChooseCitizenScreen extends StatelessWidget {
                             crossAxisCount: portrait ? 2 : 4,
                             children: snapshot.data
                                 .map<Widget>((UsernameModel user) =>
-                                CitizenAvatar(
-                                    usernameModel: user,
-                                    onPressed: () => Routes.push(context,
-                                        WeekplanSelectorScreen(user))))
+                                    CitizenAvatar(
+                                        usernameModel: user,
```
Comment (joandrsn) [inline lib/screens/choose_citizen_screen.dart]: Formatting.

File: lib/screens/edit_weekplan_screen.dart [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+    @required this.weekModel,
+    @required this.selectorBloc,
+  }) : _bloc = di.getDependency<EditWeekplanBloc>() {
+    _bloc.initializeEditBloc(user, weekModel);
+  }
+
+  /// The current week model that should be edited
+  final WeekModel weekModel;
+
+  ///
```
Comment (joandrsn) [inline lib/screens/edit_weekplan_screen.dart]: Empty comment?

**Review by joandrsn — CHANGES_REQUESTED** [review]
Overall, your linter has made MANY changes on the code. This is not good since others will have a hard time including your changes into their own code.

Please minimize only the changes to the absolute minimum.

File: lib/blocs/weekplan_selector_bloc.dart [hunk 15]
```diff
@@ -132,11 +132,33 @@ class WeekplansBloc extends BlocBase {
...
+      });
+    }
+  }
+
   /// Returns the number of marked week models
   int getNumberOfMarkedWeekModels() {
     return _markedWeekModels.value.length;
   }
 
+  // Not sure if this method breaks bloc-pattern
```
Comment (Jakobzs) [inline lib/blocs/weekplan_selector_bloc.dart:156]: It should be verified whether this method breaks the bloc pattern.

File: lib/screens/edit_weekplan_screen.dart (hunk 14 above)
Comment (Jakobzs) [inline lib/screens/edit_weekplan_screen.dart]: Empty comment.

File: lib/screens/weekplan_selector_screen.dart [hunk 16]
```diff
@@ -213,35 +216,60 @@ class WeekplanSelectorScreen extends StatelessWidget {
...
+      EditWeekPlanScreen(
+        user: _user,
+        weekModel: _weekBloc.getMarkedWeekModels()[0],
+        selectorBloc: _weekBloc,
+      ),
+    ).then((WeekModel newWeek) => _weekBloc.load(_user, true));
+    _weekBloc.toggleEditMode();
+    _weekBloc.clearMarkedWeekModels();
+
+    /*_weekBloc.markedWeekModels.listen((List<WeekModel> markedWeekModels) {
```
Comment (Jakobzs) [inline lib/screens/weekplan_selector_screen.dart]: Commented out code.

File: weekplanner.iml [hunk 17]
```diff
@@ -8,12 +8,13 @@
   <component name="NewModuleRootManager" inherit-compiler-output="true">
     <exclude-output />
     <content url="file://$MODULE_DIR$">
+      <sourceFolder url="file://$MODULE_DIR$/android/app/src/main/java" isTestSource="false" />
```
Comment (Jakobzs) [inline weekplanner.iml]: What change is this for?

File: weekplanner.iml [hunk 18]
```diff
@@ -8,12 +8,13 @@
...
       <sourceFolder url="file://$MODULE_DIR$/gen" isTestSource="false" generated="true" />
       <excludeFolder url="file://$MODULE_DIR$/.dart_tool" />
       <excludeFolder url="file://$MODULE_DIR$/.pub" />
       <excludeFolder url="file://$MODULE_DIR$/build" />
     </content>
-    <orderEntry type="jdk" jdkName="Android API 26 Platform" jdkType="Android SDK" />
+    <orderEntry type="inheritedJdk" />
```
Comment (Jakobzs) [inline weekplanner.iml]: Why is this changed?

**Review by Jakobzs — CHANGES_REQUESTED** [review]

File: lib/blocs/activity_bloc.dart (hunk 3 above)
Comment (EsbenNedergaard) [reply lib/blocs/activity_bloc.dart]: Removed the reformat that we did by mistake

File: android/app/src/main/java/io/flutter/plugins/GeneratedPluginRegistrant.java (hunk 1 above)
Comment (EsbenNedergaard) [reply android/app/src/main/java/io/flutter/plugins/GeneratedPluginRegistrant.java]: Deleted the file

File: lib/blocs/pictogram_image_bloc.dart (hunk 9 above)
Comment (EsbenNedergaard) [reply lib/blocs/pictogram_image_bloc.dart]: Done

File: lib/screens/edit_weekplan_screen.dart (hunk 14 above)
Comment (EsbenNedergaard) [reply lib/screens/edit_weekplan_screen.dart]: Fixed

File: lib/blocs/edit_weekplan_bloc.dart (hunk 4 above)
Comment (EsbenNedergaard) [reply lib/blocs/edit_weekplan_bloc.dart]: Updated the comment

File: lib/blocs/edit_weekplan_bloc.dart (hunk 6 above)
Comment (EsbenNedergaard) [reply lib/blocs/edit_weekplan_bloc.dart]: Done

File: lib/blocs/edit_weekplan_bloc.dart (hunk 8 above)
Comment (EsbenNedergaard) [reply lib/blocs/edit_weekplan_bloc.dart]: /// This method allows one to save the new information stored in the week
  /// model object and also deletes the old object if necessary

File: lib/screens/weekplan_selector_screen.dart (hunk 16 above)
Comment (EsbenNedergaard) [reply lib/screens/weekplan_selector_screen.dart]: Removed it

File: weekplanner.iml (hunk 17 above)
Comment (EsbenNedergaard) [reply weekplanner.iml]: Pretty sure this is something Intelijee did so not sure what it does

File: weekplanner.iml (hunk 18 above)
Comment (EsbenNedergaard) [reply weekplanner.iml]: Pretty sure this is something Intelijee did so not sure what it does

File: lib/blocs/new_weekplan_bloc.dart [hunk 19]
```diff
@@ -15,50 +17,65 @@ class NewWeekplanBloc extends BlocBase {
   /// The bloc contains sinks to handle when different inputs are entered and
   /// streams to tell if the inputs are valid.
   /// The bloc is also able to save the newly created weekplan.
-  /// 
+  ///
```
Comment (joandrsn) [inline lib/blocs/new_weekplan_bloc.dart:19]: Remove this if nothings changed

File: lib/blocs/new_weekplan_bloc.dart [hunk 20]
```diff
@@ -15,50 +17,65 @@ class NewWeekplanBloc extends BlocBase {
...
-      BehaviorSubject<String>();
-  final BehaviorSubject<PictogramModel> _thumbnailController =
+  @protected
+  /// This field controls the pictogram input field
+  final BehaviorSubject<PictogramModel> thumbnailController =
       BehaviorSubject<PictogramModel>();
 
   /// Handles when the entered title is changed.
-  Sink<String> get onTitleChanged => _titleController.sink;
+  Sink<String> get onTitleChanged => titleController.sink;
```
Comment (joandrsn) [inline lib/blocs/new_weekplan_bloc.dart:42]: Please dont rename variables unless you have to.

File: lib/blocs/new_weekplan_bloc.dart [hunk 21]
```diff
@@ -15,50 +17,65 @@ class NewWeekplanBloc extends BlocBase {
...
 
   /// Handles when the entered year is changed.
-  Sink<String> get onYearChanged => _yearController.sink;
+  Sink<String> get onYearChanged => yearController.sink;
```
Comment (joandrsn) [inline lib/blocs/new_weekplan_bloc.dart:45]: See above.

File: lib/blocs/weekplan_selector_bloc.dart [hunk 22]
```diff
@@ -79,7 +79,7 @@ class WeekplansBloc extends BlocBase {
           .listen((WeekModel results) {
         weekModels.add(results);
         weekModels.sort((WeekModel a, WeekModel b) {
-          if (a.name == 'Tilføj ugeplan'){
+          if (a.name == 'Tilføj ugeplan') {
```
Comment (joandrsn) [inline lib/blocs/weekplan_selector_bloc.dart]: Dont add whitespace unless needed.

**Review by joandrsn — CHANGES_REQUESTED** [review]
MUCH better!
But there are still some minor nitpicks.

File: lib/blocs/new_weekplan_bloc.dart (hunk 20 above)
Comment (EsbenNedergaard) [reply lib/blocs/new_weekplan_bloc.dart:42]: We had to because they were private otherwise and we needed access to them in a sub-class

File: lib/blocs/new_weekplan_bloc.dart (hunk 21 above)
Comment (EsbenNedergaard) [reply lib/blocs/new_weekplan_bloc.dart:45]: We had to because they were private otherwise and we needed access to them in a sub-class

File: lib/blocs/new_weekplan_bloc.dart (hunk 19 above)
Comment (joandrsn) [reply lib/blocs/new_weekplan_bloc.dart:19]: Why did you "Resolve" the comment? There are still issues.

File: lib/blocs/weekplan_selector_bloc.dart (hunk 22 above)
Comment (EsbenNedergaard) [reply lib/blocs/weekplan_selector_bloc.dart]: Done

**Review by joandrsn — APPROVED** [review]
When merging PLEASE use SQUASH! 45 commits with poor commit messages will leave an unreadable history.

**Review by Jakobzs — APPROVED** [review]

**Review by jensfors — APPROVED** [review]
